### ISSUE TITLE
wp.autograd gradient checks give wrong results for input-mutating functions and fail to compile for mixed input/output precision

### ISSUE THREAD

**Issue by daedalus5** [NVIDIA/warp#1726]
### input mutation drifts finite-difference probes

`gradcheck()`, `jacobian()`, and `jacobian_fd()` evaluate the user function repeatedly on
the same input arrays. If the function mutates an input in place — as differentiable
simulations typically do when integrating state — every evaluation after the first starts
from drifted values: the FD probes measure the wrong operating point and the check
reports spurious failures (or spurious passes).

```python
@wp.kernel
def integrate(x: wp.array[float]):
    tid = wp.tid()
    x[tid] = x[tid] + 1.0  # in-place state update

@wp.kernel
def readout(x: wp.array[float], y: wp.array[float]):
    tid = wp.tid()
    y[tid] = x[tid] * x[tid]

def forward(x):
    wp.launch(integrate, dim=4, inputs=[x])   # mutates x
    y = wp.zeros_like(x)
    wp.launch(readout, dim=4, inputs=[x], outputs=[y])
    return y

x = wp.array(np.arange(1.0, 5.0, dtype=np.float32), requires_grad=True)
wp.autograd.gradcheck(forward, inputs=[x])  # fails: each FD probe sees x incremented again
```

**Fix:** new `restore_inputs` parameter (default `True`) snapshots Warp array inputs and
restores them before every evaluation.

### mixed input/output precision breaks the FD kernel

The FD Jacobian is allocated with the *input* dtype, but the sampled output columns carry
the *output* dtype. When they differ, `compute_fd_kernel` receives e.g. a `float64`
difference with a `float32` eps and fails to compile with an input-type mismatch —
breaking `gradcheck()`, `jacobian_fd()`, and `gradcheck_tape()` for the common
"float32 parameters, float64 loss" pattern.

```python
@wp.kernel
def square_f32_to_f64(a: wp.array[wp.float32], out: wp.array[wp.float64]):
    tid = wp.tid()
    ai = wp.float64(a[tid])
    out[tid] = ai * ai

a = wp.array([1.5, 2.0], dtype=wp.float32, requires_grad=True)
out = wp.zeros(2, dtype=wp.float64, requires_grad=True)
wp.autograd.gradcheck(square_f32_to_f64, dim=2, inputs=[a], outputs=[out])
# raises a kernel compilation error on unfixed Warp
```

**Fix:** compute the difference quotient in the output precision, cast once to the
Jacobian dtype on store (also avoids quantizing eps to the narrower precision).

**Comment by shi-eric** [comment]
This is addressed by 1168a568714ffd63d3d561b326e668da5b3d5359.

The change snapshots top-level Warp array inputs passed to Python functions and restores them between gradient-check evaluations and before returning. This keeps finite-difference probes at the original operating point and prevents `jacobian()` and `jacobian_fd()` from leaving caller inputs mutated. It also computes finite-difference quotients in the output precision before casting to the input-typed Jacobian, fixing mixed-precision compilation and avoiding premature precision loss.

### Public API behavior

- `warp.autograd.gradcheck()`, `warp.autograd.jacobian()`, and `warp.autograd.jacobian_fd()` now accept `restore_inputs`, which defaults to `True` for Python functions.
- Restoration covers top-level Warp array inputs; arrays inside structs are not restored. Callers can pass `restore_inputs=False` to retain the previous behavior.
- Existing mixed-precision checks now work through `gradcheck()`, `jacobian_fd()`, and `gradcheck_tape()` when input and output arrays use different floating-point precision.

### Test coverage

- `warp/tests/test_grad_debug.py::test_gradcheck_mixed_precision` covers float32 inputs with float64 outputs and float64 inputs with float32 outputs.
- `test_gradcheck_restore_inputs` covers the reported input-mutating function, verifies that inputs remain unchanged, checks read-tracking behavior, and exercises the opt-out path.
- `test_restore_inputs_direct_jacobian` covers direct `jacobian()` and `jacobian_fd()` calls with mutated array inputs.
- `test_restore_inputs_preserves_read_state` verifies that restoration preserves caller-owned tape read state.
- `test_restore_inputs_on_exception` verifies restoration when the function or subsequent validation raises.

### Behavioral verification

The reported mutation pattern and mixed-precision behavior were also checked independently on CPU and CUDA. Input-mutating gradient checks and direct Jacobians returned the expected values without changing caller inputs, and both mixed-precision directions passed through `gradcheck()`, `jacobian_fd()`, and `gradcheck_tape()`.

### Follow-up to consider

- Add an explicit mixed-precision `gradcheck_tape()` regression test. The tape path uses the corrected finite-difference implementation and passed behavioral verification, so this does not block closure.
